lib/signex/algorithm.ex

```diff
@@ -0,0 +1,74 @@
+defmodule SignEx.Algorithm do
+  @moduledoc """
+    A `SignEx.Algorithm` struct contains all the information needed to deal with signature encryption and digest
+  """
+
+  @enforce_keys [:digest]
+  defstruct [:encryption, :digest]
+
+  defimpl String.Chars, for: SignEx.Algorithm do
+    def to_string(alg), do: "#{alg.encryption}-#{alg.digest}"
+  end
+
+  @available [
+        {:ec, :sha256},
+        {:ec, :sha384},
+        {:ec, :sha512},
+        {:rsa, :sha256},
+        {:rsa, :sha384},
+        {:rsa, :sha512},
+      ]
+  @doc """
+  List of available algorithms where each element is SignEx.Algorithm struct
+  """
+  defmacro available do
+    quote do
+      unquote(@available)
+      |> Enum.map(fn(pair) ->
+        %SignEx.Algorithm{encryption: elem(pair, 0), digest: elem(pair, 1)}
+      end)
+    end
+  end
+
+  @doc """
+  List of allowed algorithm string where each element is in format 'encryption-digest'
+  """
+  defmacro allowed_strings do
```

Comment:
I'm not sure there is any value in this being a macro. because everything is the quote block no work is done before the code is embedded.
could do something like

```
@allowed_strings = Enum.map(@available, fn({enc, digest}) -> enc <> "-" <> digest end)

def allowed_strings, do: @allowed_strings
```